### DISCUSSION TITLE
Nano Digi USB support

### DISCUSSION THREAD

**Discussion by Martin-By** [mrene/minidsp-rs#149] [General]
Hi,
I am not profession in linux coding, but could install the executable files within /usr/local/bin.
Running "minidsp" it returns: "Error: No devices found"

The Device is found by system:
lsusb:  Bus 003 Device 002: ID 04d8:003f Microchip Technology, Inc. nanoDIGI v2.6

and mapped as hidraw0:
crw-rw-rw- 1 root root 242, 0 Jun 13 11:09 /dev/hidraw0

within "/etc/udev/rules.d" I created a file "50-minidsp_usb.rules"
SUBSYSTEM=="usb", ATTR{idVendor}=="04d8", ATTR{idProduct}=="003f", MODE="0666"

ubuntu@musicserver1:~$ cat /sys/class/hidraw/hidraw0/device/uevent
DRIVER=hid-generic
HID_ID=0003:000004D8:0000003F
HID_NAME=miniDSP LTD. nanoDIGI v2.6
HID_PHYS=usb-0000:00:1d.1-1/input0
HID_UNIQ=
MODALIAS=hid:b0003g0001v000004D8p0000003F

root@musicserver1:/dev/input# minidsp probe
thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 2', protocol/src/commands.rs:575:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

What I am doing wrong with my setup to use minidsp?

My Device: MiniDSP_2x8_nanoDIGI_204
 Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-74-generic x86_64)

**Comment by mrene** [comment]
Hi! I think you're the first one to try it out with a nano-digi!

At first glance it seems that there is a slight difference in how the device responds with it's hardware id - could you try this: `minidsp -vvv probe` - it would still crash but I'm hoping it'll print some packets before :)

A successful run would start with:
```
Sent: [31]
Sent: ReadHardwareId
Recv: [31, 01, 0a, 0a]
Recv: HardwareId { payload: [01, 0a, 0a] }
```

Alternatively, `RUST_LOG="minidsp=trace" minidsp probe` would also yield something comparable.
```
[2021-06-16T21:35:42Z TRACE minidsp::transport::multiplexer] send: ReadHardwareId
[2021-06-16T21:35:42Z TRACE minidsp::transport::hid::stream] read: [05, 31, 01, 0a, 0a, ff, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 00, 04]
[2021-06-16T21:35:42Z TRACE minidsp::transport::multiplexer] recv: HardwareId { payload: [01, 0a, 0a] }
```

**Comment by Martin-By** [comment]
So now the result for my NanoDigi device:

root@musicserver1:/etc/environment.d# RUST_LOG="minidsp=trace" minidsp probe
[2021-06-17T08:39:08Z TRACE minidsp::transport::multiplexer] send: ReadHardwareId
[2021-06-17T08:39:08Z TRACE minidsp::transport::hid::stream] read: [04, 31, 02, 09, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, 02]
[2021-06-17T08:39:08Z TRACE minidsp::transport::multiplexer] recv: HardwareId { payload: [02, 09] }

ubuntu@musicserver1:~/.local/share$ sudo minidsp -vvv probe
[sudo] password for ubuntu:
Sent: [31]
Sent: ReadHardwareId
Recv: [31, 02, 09]
Recv: HardwareId { payload: [02, 09] }
thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 2', protocol/src/commands.rs:575:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by mrene** [comment]
Thanks, I made a tentative fix, can you [try this build](https://github.com/mrene/minidsp-rs/actions/runs/946340137) and let me know if it works?

**Reply by Martin-By** [reply]
Hi, thank you for this quick fix!

I installed the deb-file:

ubuntu@musicserver1:/usr/local/bin$ sudo service minidsp status
● minidsp.service - MiniDSP Server
     Loaded: loaded (/lib/systemd/system/minidsp.service; enabled; vendor preset: enabled)
     Active: inactive (dead)

sudo minidsp
MasterStatus { preset: 0, source: NotInstalled, volume: Gain(-5.0), mute: false, dirac: false }
Input levels:
Output levels:

ubuntu@musicserver1:/$ sudo minidsp probe
Found Generic with serial 900260 at usb:0003%3A0002%3A00?vid=04d8&pid=003f [hw_id: 9, dsp_version: 54]

After a reboot:

ubuntu@musicserver1:~$ sudo service minidsp status
[sudo] password for ubuntu:
● minidsp.service - MiniDSP Server
     Loaded: loaded (/lib/systemd/system/minidsp.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2021-06-17 17:08:52 CEST; 1min 4s ago
   Main PID: 840 (minidspd)
      Tasks: 8 (limit: 4595)
     Memory: 8.4M
     CGroup: /system.slice/minidsp.service
             └─840 /usr/bin/minidspd --config /etc/minidsp/config.toml

Jun 17 17:08:52 musicserver1 systemd[1]: Started MiniDSP Server.
Jun 17 17:08:52 musicserver1 minidspd[840]: App is running on: 127.0.0.1:5380
Jun 17 17:08:52 musicserver1 minidspd[840]: App is listening on: /tmp/minidsp.sock

Trying to connect to the server on Port 5380 on my local IP it deny the connection!
Within /etc/minidsp/config.toml I canged from 127.0.0.0:port into my local IP of machine. is this OK?

http://mylocalIP:5380/devices 
[
  {
    "url": "usb:0003%3A0002%3A00?vid=04d8&pid=003f",
    "version": {
      "hw_id": 9,
      "dsp_version": 54,
      "serial": 900260
    },
    "product_name": "Generic"
  }
]

So far it looks good! So can I test some commands now?

"minidsp gain -- -30" results in:
thread 'thread 'tokio-runtime-workertokio-runtime-worker' panicked at '' panicked at 'range start must not be greater than end: 1 <= 0range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:', 245/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs::9245
:note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
9


ubuntu@musicserver1:~$ RUST_BACKTRACE=full minidsp gain -- -30
thread 'tokio-runtime-worker' panicked at 'range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rsthread ':tokio-runtime-worker245' panicked at ':range start must not be greater than end: 1 <= 09',
/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:245stack backtrace:
:9
   0:     0x561879ed4e64 - <unknown>
   1:     0x561879d8b46f - <unknown>
   2:     0x561879ed45f1 - <unknown>
   3:     0x561879ed3ff5 - <unknown>
   4:     0x561879ed2e2a - <unknown>
   5:     0x561879ef0a38 - <unknown>
   6:     0x561879ef09ac - <unknown>
   7:     0x561879ef095d - <unknown>
   8:     0x561879be3870 - <unknown>
   9:     0x561879e5048f - <unknown>
  10:     0x561879e212d6 - <unknown>
  11:     0x561879e0e676 - <unknown>
  12:     0x561879e10c48 - <unknown>
  13:     0x561879f129a0 - <unknown>
  14:     0x561879f11114 - <unknown>
  15:     0x561879f10203 - <unknown>
  16:     0x561879f04b06 - <unknown>
  17:     0x561879f0478e - <unknown>
  18:     0x561879efbb3a - <unknown>
  19:     0x7f6779d3a609 - start_thread
  20:     0x7f6779b0a293 - clone
  21:                0x0 - <unknown>
stack backtrace:
   0:     0x561879ed4e64 - <unknown>
   1:     0x561879d8b46f - <unknown>
   2:     0x561879ed45f1 - <unknown>
   3:     0x561879ed3ff5 - <unknown>
   4:     0x561879ed2e2a - <unknown>
   5:     0x561879ef0a38 - <unknown>
   6:     0x561879ef09ac - <unknown>
   7:     0x561879ef095d - <unknown>
   8:     0x561879be3870 - <unknown>
   9:     0x561879e5048f - <unknown>
  10:     0x561879bf2d96 - <unknown>
  11:     0x561879e0e676 - <unknown>
  12:     0x561879e10c48 - <unknown>
  13:     0x561879f12973 - <unknown>
  14:     0x561879f113d2 - <unknown>
  15:     0x561879f10203 - <unknown>
  16:     0x561879f04b06 - <unknown>
  17:     0x561879f0478e - <unknown>
  18:     0x561879efbb3a - <unknown>
  19:     0x7f6779d3a609 - start_thread
  20:     0x7f6779b0a293 - clone
  21:                0x0 - <unknown>

**Reply by mrene** [reply]
Could you run the last cmd with some tracing? `RUST_LOG="minidsp=trace" minidsp gain -- -30`

**Reply by Martin-By** [reply]
ubuntu@musicserver1:~$ RUST_LOG="minidsp=trace" minidsp gain -- -30
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] send: ReadHardwareId
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] recv: HardwareId { payload: [02, 09] }
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] send: ReadMemory { addr: ffa1, size: 01 }
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] recv: MemoryData(MemoryView { base: ffa1, data: [36] })
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] send: ReadMemory { addr: fffc, size: 02 }
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] recv: MemoryData(MemoryView { base: fffc, data: [01, 04] })
[2021-06-18T05:05:48Z TRACE minidsp::transport::multiplexer] send: SetVolume { value: Gain(-30.0) }
thread 'tokio-runtime-worker' panicked at 'range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:245:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker' panicked at 'range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:245:9

**Reply by mrene** [reply]
Can you try to stop the service first so the device is queried directly? `systemctl stop minidsp.service`

**Reply by Martin-By** [reply]
Hi, Thanks!
I understand:
ubuntu@musicserver1:~$ sudo systemctl stop minidsp.service
[sudo] password for ubuntu:
ubuntu@musicserver1:~$ sudo RUST_LOG="minidsp=trace" minidsp gain -- -30
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] send: ReadHardwareId
[2021-06-18T12:53:58Z TRACE minidsp::transport::hid::stream] read: [04, 31, 02, 09, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, 74]
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] recv: HardwareId { payload: [02, 09] }
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] send: ReadMemory { addr: ffa1, size: 01 }
[2021-06-18T12:53:58Z TRACE minidsp::transport::hid::stream] read: [05, 05, ff, a1, 36, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, 75]
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] recv: MemoryData(MemoryView { base: ffa1, data: [36] })
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] send: ReadMemory { addr: fffc, size: 02 }
[2021-06-18T12:53:58Z TRACE minidsp::transport::hid::stream] read: [06, 05, ff, fc, 01, 04, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, 76]
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] recv: MemoryData(MemoryView { base: fffc, data: [01, 04] })
[2021-06-18T12:53:58Z TRACE minidsp::transport::multiplexer] send: SetVolume { value: Gain(-30.0) }
[2021-06-18T12:53:58Z TRACE minidsp::transport::hid::stream] read: [00, 42, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, ff, 77]
thread 'tokio-runtime-worker' panicked at 'range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:245:9thread '
tokio-runtime-workernote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
' panicked at 'range start must not be greater than end: 1 <= 0', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/bytes-1.0.1/src/bytes.rs:245:9

**Reply by mrene** [reply]
That's odd, do you know if it actually set the volume? Does `minidsp` report the new value?

**Reply by Martin-By** [reply]
No, minidsp does not report the new value. Furthermore there seem to be a difference running it with sudo or not (with service stopped)
Rrunning without sudo:
ubuntu@musicserver1:~$ minidsp
Error: No devices found

with sudo:
ubuntu@musicserver1:~$ sudo minidsp
MasterStatus { preset: 0, source: NotInstalled, volume: Gain(-5.0), mute: false, dirac: false }
Input levels:
Output levels:

**Reply by mrene** [reply]
Ah there is probably a udev permission missing for running as non-root. When the service is present, it always goes through it so you don't need to have raw device access.

Now, it looks like the device isn't accepting this command at all 🤔 Do other commands work? Can you turn mute on-off with `minidsp -vvv mute on` ?

If these don't work, you can run `minidsp -vvv server` and use the official plugin to connect (using the ip field) - you should see the commands sent as you do changes in the plugin. Paste the commands used for master volume and we can compare them.

**Reply by Martin-By** [reply]
Hi, yes the verbose mute call works. So far it seem to work in principle.
But I can not setup the connection with the MiniDSP-Plugin to collect commands.
The regular service is available do to service running  "LocalIP:5380/devices". 
With that running MiniDSP-Plugin I entered "LocalIP:5380" right?
Trying to connect via server does not work: neither service server nor service stopped but verbose server started.

What I am doing wrong?

**Reply by Martin-By** [reply]
Amazing:
running the service minidsp and config the advertisement of device:
192.168.2044:5380/devices:
[
  {
    "url": "tcp://192.168.20.44:5333?name=Musicserver1",
    "version": null,
    "product_name": null
  },
  {
    "url": "usb:0003%3A0002%3A00?vid=04d8&pid=003f",
    "version": {
      "hw_id": 9,
      "dsp_version": 54,
      "serial": 900260
    },
    "product_name": "Generic"
  }
]

Now on PC with windows using the MIniDSP App and using the advertized name "Musicserver1" I can access all things.

So how to run the non service minidsp server to get the command strings visible?

**Reply by Martin-By** [reply]
So far I did following

- reboot
- stop the service
- as su user (root)
-  RUST_LOG="minidsp=trace" minidsp -vvv server

See attached file:
change channel: OK
change config: OK
load configs into cannel 1 & 2: OK
Mute:OK
Mute Channel 1 Main Level: error and Server stops
[2021-06_18_minidsp_trace_server.txt](https://github.com/mrene/minidsp-rs/files/6680333/2021-06_18_minidsp_trace_server.txt)
`

After restarting the service
Muting of channel 1 is OK!

Perhaps the Config file of  Config 1 helps (see attached file)

[2015_04_MiniDSP-2x8-nanoDIGI_setting.zip](https://github.com/mrene/minidsp-rs/files/6680373/2015_04_MiniDSP-2x8-nanoDIGI_setting.zip)

**Reply by mrene** [reply]
Thanks!

By default the service is only configured to listen on 127.0.0.1, you need to change the bind addresses to be able to use it from another machine. Check the commented lines here:

https://github.com/mrene/minidsp-rs/blob/3c301d47883cf3f03ec483b152e90eee76c53836/docs/config.example.toml#L7-L16

The logs you provided are in the right format, but I don't see any change to the master volume (restoring a confiuguration doesn't use the same commands as changing the values, unfortunately).  Can you change the master volume and show me what command it sends?

The config file will help if we want to control all settings, then I can show you how to map each parameter.

**Reply by Martin-By** [reply]
Hi, I do not see any option to change the gain, I only see the setting within a gray field right between IP Address field and green Update button. I can only set the both channels gain not for master volume.

I already set within the /etc/minidsp/config.toml file (for the server as service).
Is there any other location for the config file for the non service server (minidsp -vvv server)?

**Reply by Martin-By** [reply]
my /etc/minidsp/config.toml
[config.toml.txt](https://github.com/mrene/minidsp-rs/files/6680709/config.toml.txt)
![NanoDigi2x8](https://user-images.githubusercontent.com/21987914/122643996-131a3300-d113-11eb-8bfa-bda96678f361.jpg)

**Reply by mrene** [reply]
Can you click on the grayed out area and change the value? Although if it's not possible to change it, it would explain why the command fails.

There is no config file for `minidsp server`, you should be able to enter your IP into the ip address field in the plugin and just connect `ifconfig` should tell you what it is.

**Reply by Martin-By** [reply]
No, the greyed field is an inactive display. This board has an IR remote device connection! With this I can control the master sound level, so this command seems to be reserved to IR remote only!?

The difference I can not understand:
Running minidsp server as service does not raise  an error for one channel Mute or editing the crossover variables.

**Reply by mrene** [reply]
And running the "normal" service consistently errors when changing settings ?

**Reply by Martin-By** [reply]
Running the server as normal service (see my config above attached) with "advertise = { ip = "192.168.2.44", name="Musicserver1" }" and connection by name, I can change all parameters without any crash (so far I tested criss cross).
Can I start this server with verbose parameters to collect the commands comming from MiniDSP APP?

**Reply by mrene** [reply]
I'm surprised that the other cmd crashes since it runs the same logic, you can use `minidspd -vvv --log commands.txt -c  /etc/minidsp/config.toml` to start the service manually and log commands with some debug output. `commands.txt` will just contain the hex packets and the protocol debug will be printed to stderr.

Oh thinking further it might be the protocol decoder enabled with `-vvv` that causes the crash. Getting things with `--log` should allow me to reproduce the issue. So I guess: try it with `-vvv --log file.txt` and with `--log file.txt` to see if the former crashes or not.

**Reply by Martin-By** [reply]
Yes, I will try this tommorw morning.

**Reply by mrene** [reply]
I fixed the protocol decoder, and added your xml config to be able to decode what's being sent - [grab the latest build](https://github.com/mrene/minidsp-rs/actions/runs/953216040). It also has updated udev rules so you shouldn't need to `sudo` anymore.
 
You should be able to view the commands being set for the input gain + mute. From what you've shown me, the input 1 mute would be: 
`Sent: Write { addr: 35, value: Value { Float: 0.0 (Bytes: [0, 0, 0, 0]) }, addr_len: 01 } (0x35 | 53) <> MUTE_NO_SLEW_ALG_7_1MUTE`

The string `MUTE_NO_SLEW_ALG_7_1MUTE` is what we're after, there will be a different one for each control.

If you want to retrieve the addresses corresponding to each control, I can add them so the nanodigi is fully supported by minidsp-rs.

**Reply by Martin-By** [reply]
Hello Mathieu, thank you very much for this new version.

Running the server by service (without verbose) all seem to work as expected. e.g. I can enter the outputchannel of my crossover and can mute any of them wihtout any crash.

Furthermore I think you can implement the master gain command again. Let us try. Very late I realized that there are some Android out there to run MiniDSP commands. They use the Master Gain.

But when stopping the server and start minidspd with verbose option as you mentioned it crashes again when using some channel dependend command. Perhaps there is still a bug in the verbose module (I think).

**Reply by mrene** [reply]
That's possible, can you show me the logs up until the crash point?

**Reply by Martin-By** [reply]
Hi,
here some files:

the commands.txt
[commands.txt](https://github.com/mrene/minidsp-rs/files/6682624/commands.txt)

the command line output:
[commands_until_crash.txt](https://github.com/mrene/minidsp-rs/files/6682625/commands_until_crash.txt)

the last command before crash (trying bypass):
![Last_command_when_crash](https://user-images.githubusercontent.com/21987914/122680348-f4896a00-d1ee-11eb-8962-536ee0a1e419.jpg)

**Reply by mrene** [reply]
Ah it appears that the biquad commands use 1 byte addresses instead of 2 as well - that'll require some more adjustments.

**Reply by mrene** [reply]
Can you change the PEQ settings and show me the commands.txt file? (remove `-vvv` so it doesn't crash)

In fact, it'd help if you could change the input mute and gain as well just so I can see the commands.

**Reply by mrene** [reply]
Oh and click on `monitor` to see what it looks at to get the input levels.

**Reply by Martin-By** [reply]
Hi, I hope all is recorded, so you can interpret the commands (without -vvvv the screen output did not show the used commands as with vervobe option).
I used as much of the available commands. Perhaps we need more adjustments later. 
[commands2.txt](https://github.com/mrene/minidsp-rs/files/6682704/commands2.txt)

**Reply by Martin-By** [reply]
Here my control panel, I used channel 1,2 & 7,8
![Screenshot 2021-06-20 183709](https://user-images.githubusercontent.com/21987914/122681864-abd5af00-d1f6-11eb-82d8-821133e252bb.jpg)

**Reply by Martin-By** [reply]
Hi, today I used an Android App to use the Master Gain to create a Command File:
[commands3.txt](https://github.com/mrene/minidsp-rs/files/6688688/commands3.txt)

I hope there is some useful code available from this.
Best regards, and thank you for this piece of code, I enjoy!
Martin

**Reply by mrene** [reply]
Ah, this confirms the command being sent by minidsp-rs is already the right one, and the werid device response should not be considered as an error but as a different kind of acknowledgement.

**Reply by Martin-By** [reply]
The send command for Master Gain by Android APP did not change Level of Master Gain on the minidsp board.

**Reply by mrene** [reply]
Oh that is unfortunate :(

**Reply by Martin-By** [reply]
Yeah, I forgot to make clear in my earlier post, sorry for that.
